### PR TITLE
feat: Multivariate Anomaly Detection

### PR DESCRIPTION
**Pull request by ab93**
Additions:
- Beta VAE to balance the KLD and Log likelihood loss
- RobustMahalanobis Distance for multivariate thresholding
- Percentile based thresholding with aggregation
- Allow a tag suffix in docker workflow when using the github UI
- Exponential moving average in postprocess output
- Introduce pytests instead of unittests for newer test cases
- Improve synthetic anomaly generator module
- Per feature anomaly score when using certain thresholding models
- Score config for different feature and window aggreation methods

Removal:
- backtesting cli tool (since this will be added in a separate backtesting tool UI)

### CODE REVIEW

File: numalogic/udfs/tools.py [hunk 1]
```diff
@@ -250,7 +268,7 @@ def __construct_train_key(keys: KEYS) -> str:
     def __fetch_ts(self, key: str) -> _DedupMetadata:
         try:
             data = self.client.hgetall(key)
-        except RedisError:
+        except Exception:
```
Comment (s0nicboOm) [inline numalogic/udfs/tools.py:271]: Any specific reason for this?

File: numalogic/udfs/postprocess.py [hunk 2]
```diff
@@ -197,7 +198,7 @@ def _per_feature_score(feat_names: list[str], scores: NDArray[float]) -> dict[st
     @classmethod
     def compute(
         cls, model: artifact_t, input_: NDArray[float], postproc_clf=None, **_
-    ) -> NDArray[float]:
+    ) -> tuple[NDArray[float], NDArray[float]]:
```
Comment (s0nicboOm) [inline numalogic/udfs/postprocess.py:201]: I thought we said we donot want to return 2 things when working on this.

File: numalogic/udfs/postprocess.py [hunk 3]
```diff
@@ -29,6 +30,7 @@
 LOCAL_CACHE_TTL = int(os.getenv("LOCAL_CACHE_TTL", "3600"))
 LOCAL_CACHE_SIZE = int(os.getenv("LOCAL_CACHE_SIZE", "10000"))
 LOAD_LATEST = os.getenv("LOAD_LATEST", "false").lower() == "true"
+EXP_MOV_AVG_BETA = float(os.getenv("EXP_MOV_AVG_BETA", "0.6"))
```
Comment (s0nicboOm) [inline numalogic/udfs/postprocess.py:33]: can we add a TODO here please

File: numalogic/transforms/_stateless.py [hunk 4]
```diff
@@ -53,3 +59,76 @@ def transform(self, X, **__):
...
+
+    def transform(self, x: npt.NDArray[float], **__) -> npt.NDArray[float]:
+        noise = self._rng.normal(loc=0.0, scale=self._scale, size=x.shape)
+        if self._is_abs:
+            noise = np.abs(noise)
+        return x + noise
+
+
+class DifferenceTransform(StatelessTransformer):
+    def transform(self, input_: npt.NDArray, **__):
```
Comment (s0nicboOm) [inline numalogic/transforms/_stateless.py:132]: knipick: can we have the docstrings added

File: numalogic/tools/trainer.py [hunk 5]
```diff
@@ -80,5 +80,5 @@ def predict(self, model: Optional[LightningModule] = None, unbatch=True, **kwarg
         recon_err = super().predict(model, **kwargs)
         recon_err = torch.vstack(recon_err)
         if unbatch:
-            return inverse_window(recon_err, method="keep_last")
+            return inverse_window(recon_err, method="keep_first")
```
Comment (s0nicboOm) [inline numalogic/tools/trainer.py:83]: any specific reason for this?

File: numalogic/models/threshold/_median.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+import numpy as np
+import numpy.typing as npt
+from typing_extensions import Self, Final
+
+from numalogic.base import BaseThresholdModel
+from numalogic.tools.exceptions import InvalidDataShapeError, ModelInitializationError
+
+_INLIER: Final[int] = 0
+_OUTLIER: Final[int] = 1
```
Comment (s0nicboOm) [inline numalogic/models/threshold/_median.py:12]: I think we should add this to config too. Like we have for min_threshold

**Review by s0nicboOm — COMMENTED** [review]
Some comments

File: numalogic/udfs/postprocess.py (hunk 2 above)
Comment (ab93) [reply numalogic/udfs/postprocess.py:201]: Yes as discussed, will change it.

File: numalogic/udfs/postprocess.py (hunk 3 above)
Comment (ab93) [reply numalogic/udfs/postprocess.py:33]: Probably won't need this with the new config

File: numalogic/tools/trainer.py (hunk 5 above)
Comment (ab93) [reply numalogic/tools/trainer.py:83]: Better reconstruction if we `keep_first`

File: numalogic/models/threshold/_median.py (hunk 6 above)
Comment (ab93) [reply numalogic/models/threshold/_median.py:12]: Not really. It is used in the `predict` method for which the output is just Boolean. You would never want anything else here.

File: numalogic/udfs/tools.py (hunk 1 above)
Comment (ab93) [reply numalogic/udfs/tools.py:271]: Yes there is an importError that happens when the user does not install Redis as an optional dependency.

File: numalogic/tools/aggregators.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import os
+from collections.abc import Callable
+
+import numpy as np
+import numpy.typing as npt
+
+from numalogic.transforms import expmov_avg_aggregator
+
+
+EXP_MOV_AVG_BETA = float(os.getenv("EXP_MOV_AVG_BETA", "0.6"))
```
Comment (s0nicboOm) [inline numalogic/tools/aggregators.py:10]: this will come via the conf now. Right?

File: numalogic/udfs/postprocess.py [hunk 8]
```diff
@@ -137,11 +127,17 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
...
         if payload.status in (Status.ARTIFACT_FOUND, Status.ARTIFACT_STALE) and thresh_artifact:
             try:
-                anomaly_scores = self.compute(
+                feature_scores = self.compute(
                     model=thresh_artifact.artifact,
                     input_=payload.get_data(),
+                    score_conf=_conf.numalogic_conf.score,
                     postproc_clf=postproc_clf,
+                )  # (nfeat,)
+
```
Comment (s0nicboOm) [inline numalogic/udfs/postprocess.py:136]: Can we add `_update_info_metric`. Similar to like https://github.com/numaproj/numalogic/blob/780f2e23cccf3648b385499c589347c3bed0461e/numalogic/udfs/inference.py#L162C13-L162C32

File: numalogic/udfs/postprocess.py [hunk 9]
```diff
@@ -206,39 +196,75 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
 
     @staticmethod
     def _per_feature_score(feat_names: list[str], scores: NDArray[float]) -> dict[str, float]:
-        if scores.shape[0] == 1:
-            # TODO improve this to incorporate per feature anomaly insights
-            return {_name: scores.item() for _name in feat_names}
+        if len(scores) != len(feat_names):
+            return {}
```
Comment (s0nicboOm) [inline numalogic/udfs/postprocess.py:200]: can we also add a warn log here

File: numalogic/config/_config.py [hunk 10]
```diff
@@ -98,6 +99,34 @@ class TrainerConf:
...
+
+
+@dataclass
+class AggregatorConf:
+    method: AggMethod
+    conf: dict[str, Any] = field(default_factory=dict)
+
+
+@dataclass
+class ScoreConf:
```
Comment (s0nicboOm) [inline numalogic/config/_config.py:121]: Should not this come under udf config?

**Review by s0nicboOm — COMMENTED** [review]
minor comments

File: numalogic/tools/aggregators.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/tools/aggregators.py:10]: Correct. Removing.

File: numalogic/udfs/postprocess.py (hunk 8 above)
Comment (ab93) [reply numalogic/udfs/postprocess.py:136]: I will leave that one for you to take care . :)

File: numalogic/config/_config.py (hunk 10 above)
Comment (ab93) [reply numalogic/config/_config.py:121]: Not really, aggregation is needed for the end score. I believe it should come under the base numalogic conf.

File: tests/transforms/test_transforms.py [hunk 11]
```diff
@@ -1,88 +1,152 @@
...
 
 import numpy as np
-from numpy.testing import assert_almost_equal, assert_array_less
+import pytest
+from numpy.testing import assert_almost_equal, assert_array_less, assert_array_equal
 from sklearn.pipeline import make_pipeline
 
 from numalogic.base import StatelessTransformer
-from numalogic.transforms import LogTransformer, StaticPowerTransformer, TanhScaler, FlattenVector
+from numalogic.transforms import (
```
Comment (qhuai) [inline tests/transforms/test_transforms.py:9]: Which transformers in the list are currently used? The info will help us prioritize which part to code review now.

**Review by qhuai — COMMENTED** [review]
Share the current set of comments. Still working through files in https://github.com/numaproj/numalogic/blob/main/numalogic/models/vae

File: numalogic/transforms/_stateless.py [hunk 12]
```diff
@@ -55,6 +61,84 @@ def inverse_transform(self, X) -> npt.NDArray[float]:
...
+    def transform(self, x: npt.NDArray[float], **__) -> npt.NDArray[float]:
+        _df = pd.DataFrame(x, dtype=np.float32)
+        if (self.lower is not None) and (self.upper is not None):
+            return _df.clip(lower=self.lower, upper=self.upper, axis=1).to_numpy(dtype=np.float32)
+        if self.upper is not None:
+            return _df.clip(upper=self.upper, axis=1).to_numpy(dtype=np.float32)
+        return _df.clip(lower=self.lower, axis=1).to_numpy(dtype=np.float32)
+
+
+class GaussianNoiseAdder(StatelessTransformer):
```
Comment (qhuai) [inline numalogic/transforms/_stateless.py:104]: This is probably a redundant comment. I don't find where this class is used, except in its unit test.

**Review by qhuai — COMMENTED** [review]
Have you evaluated the quality difference between plain VAE v.s. conv1d VAE in anomaly detection?

File: numalogic/models/threshold/_median.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        y_hat = np.zeros(x.shape, dtype=int)
+        y_hat[x > self._thresh] = _OUTLIER
+        return y_hat
+
+    def score_samples(self, x: npt.NDArray[float]) -> npt.NDArray[float]:
+        if not self._is_fitted:
+            raise ModelInitializationError("Model not fitted yet.")
+
+        self._validate_input(x)
+        return x / self._thresh
```
[the remark this answers is not in the document]
Comment (s0nicboOm) [reply numalogic/models/threshold/_median.py:68]: I see where @qhuai is coming from but this will only happen if user explicitly sets min_threshold to 0.I think we should set a condition for min_threshold can never be 0.

File: numalogic/models/threshold/_median.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
+import numpy as np
+import numpy.typing as npt
+from typing_extensions import Self, Final
+
+from numalogic.base import BaseThresholdModel
+from numalogic.tools.exceptions import InvalidDataShapeError, ModelInitializationError
+
+_INLIER: Final[int] = 0
+_OUTLIER: Final[int] = 1
+_INPUT_DIMS: Final[int] = 2
```
Comment (s0nicboOm) [inline numalogic/models/threshold/_median.py:10]: You think these can be class variables or another ENUM class?

Comment (s0nicboOm) [reply numalogic/models/threshold/_median.py:10]: I think we should make it a ENUM class because even _std.py can use that.

File: numalogic/udfs/factory.py [hunk 15]
```diff
@@ -27,6 +26,8 @@ class UDFFactory:
     from numalogic.udfs.preprocess import PreprocessUDF
     from numalogic.udfs.trainer import DruidTrainerUDF, PromTrainerUDF
 
+    nl_udf_t = TypeVar("nl_udf_t", bound=NumalogicUDF, covariant=True)
```
Comment (s0nicboOm) [inline numalogic/udfs/factory.py:29]: Covariant =True not needed. It should be strictly bounded to NumalogicUDF.

**Review by s0nicboOm — COMMENTED** [review]
Some more minor comments.

**Review by s0nicboOm — APPROVED** [review]
Also, I am not sure how do we test pytest files. can you add a step in README for it too.

Added some minor comments otherwise LGTM.

File: numalogic/models/vae/variants/conv.py [hunk 16]
```diff
@@ -101,7 +99,7 @@ def forward(self, x: Tensor) -> tuple[Tensor, Tensor]:
 
 class Decoder(nn.Module):
     """
-    Decoder (non-probabilistic) module for Convolutional Variational Autoencoder.
+    Decoder module for Convolutional Variational Autoencoder.
```
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/models/vae/variants/conv.py:102]: No, it is not equivalent. Padding is mostly done with zeros in Conv layers. Padding is done so that the values at the edge of the sequence have the same weightage as the non-edged data points.

File: numalogic/transforms/_stateless.py (hunk 12 above)
Comment (ab93) [reply numalogic/transforms/_stateless.py:104]: Called from config

File: numalogic/models/threshold/_median.py (hunk 14 above)
Comment (ab93) [reply numalogic/models/threshold/_median.py:10]: We can do that later. Minor thing.

File: numalogic/udfs/factory.py (hunk 15 above)
Comment (ab93) [reply numalogic/udfs/factory.py:29]: Not really. For example PostprocessUDF has multi compute functions. We need the linter to recognize the child classes of NumalogicUDF

File: numalogic/models/threshold/_median.py [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class MaxPercentileThreshold(BaseThresholdModel):
+    """
+    Percentile based Thresholding estimator.
+
+    Args:
+        max_inlier_percentile: Max percentile greater than which will be treated as outlier
+        min_threshold:  Value to be used if threshold is less than this
+        aggregate: If True, combine the per-feature scores into 1
```
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/models/threshold/_median.py:20]: Good catch. Fixing.

File: tests/transforms/test_transforms.py (hunk 11 above)
Comment (ab93) [reply tests/transforms/test_transforms.py:9]: Just the changed files.

**Comment by ab93** [comment]
> Have you evaluated the quality difference between plain VAE v.s. conv1d VAE in anomaly detection?

Yes, conv1d has much better quantitative scores.

**Comment by ab93** [comment]
> Also, I am not sure how do we test pytest files. can you add a step in README for it too.
> 
> Added some minor comments otherwise LGTM.

Simple `make test` will work, already in README.

File: numalogic/models/threshold/_median.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+        self._is_fitted = True
+        return self
+
+    def predict(self, x: npt.NDArray[float]) -> npt.NDArray[int]:
+        if not self._is_fitted:
+            raise ModelInitializationError("Model not fitted yet.")
+        self._validate_input(x)
+
+        y_hat = np.zeros(x.shape, dtype=int)
+        y_hat[x > self._thresh] = _OUTLIER
```
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/models/threshold/_median.py:58]: There is no hidden assumption. We are just initializing `y_hat`. Feel free to do some numpy experiments and read about [boolean array indexing](https://numpy.org/doc/stable/user/basics.indexing.html#boolean-array-indexing), and let me know if you have any questions.
